File: edge/worker/src/store.ts
/**
 * fox-shield edge worker — persistence layer.
 *
 * A Store abstraction backed by Cloudflare KV (SHIELD_KV) with an in-memory
 * fallback for local development when the KV binding is not present. The key
 * schema is shared with the Go origin shield:
 *
 *   ratelimit:{ip}   -> sliding-window request timestamps (JSON array of ms)
 *   ban:{ip}         -> banned IPs (value = reason)
 *   dark:{hash}      -> dark-listed request hashes (value = normalized request)
 *   destroy:{ip}     -> destroyed-request counter (value = count)
 *
 * KVNamespace is the Cloudflare Workers KV binding type. When it is undefined
 * (local dev / tests) the MemoryStore is used so the worker still runs.
 */

/** Minimal KV surface we depend on, so tests can inject a fake. */
export interface KVLike {
  get(key: string): Promise<string | null>;
  put(key: string, value: string, options?: { expirationTtl?: number }): Promise<void>;
  delete(key: string): Promise<void>;
}

/** The Cloudflare KVNamespace binding type (from @cloudflare/workers-types). */
export type KVNamespace = KVLike;

/** Store contract used by the shield middleware chain. */
export interface Store {
  /** Returns the value under key, or null when absent/expired. */
  get(key: string): Promise<string | null>;
  /** Stores value under key with an optional TTL in seconds (0 = no expiry). */
  set(key: string, value: string, ttlSeconds?: number): Promise<void>;
  /** Deletes a key. */
  delete(key: string): Promise<void>;
  /** Returns all non-expired values (used by the similarity dark-list scan). */
  snapshot(): Promise<string[]>;
}

/** Key helpers centralize the shared schema (mirrors internal/store/keys.go). */
export function rateLimitKey(ip: string): string {
  return `ratelimit:${ip}`;
}

export function banKey(ip: string): string {
  return `ban:${ip}`;
}

export function darkKey(hash: string): string {
  return `dark:${hash}`;
}

export function destroyKey(ip: string): string {
  return `destroy:${ip}`;
}

/** KV-backed Store. */
export class KVStore implements Store {
  private readonly kv: KVNamespace;

  constructor(kv: KVNamespace) {
    this.kv = kv;
  }

  async get(key: string): Promise<string | null> {
    return this.kv.get(key);
  }

  async set(key: string, value: string, ttlSeconds?: number): Promise<void> {
    if (ttlSeconds && ttlSeconds > 0) {
      await this.kv.put(key, value, { expirationTtl: ttlSeconds });
    } else {
      await this.kv.put(key, value);
    }
  }

  async delete(key: string): Promise<void> {
    await this.kv.delete(key);
  }

  async snapshot(): Promise<string[]> {
    // KV does not support enumeration from the Worker runtime. The similarity
    // detector therefore relies on the in-memory fallback for the dark-list
    // scan; with a real KV binding, exact dark-list hits still work via get().
    return [];
  }
}

interface MemItem {
  value: string;
  expiresAt: number; // epoch ms, 0 = no expiry
}

/** Thread-safe in-memory Store used when KV is not bound (local dev / tests). */
export class MemoryStore implements Store {
  private readonly items = new Map<string, MemItem>();

  private isExpired(item: MemItem, now: number): boolean {
    return item.expiresAt !== 0 && now >= item.expiresAt;
  }

  async get(key: string): Promise<string | null> {
    const item = this.items.get(key);
    if (!item) {
      return null;
    }
    if (this.isExpired(item, Date.now())) {
      this.items.delete(key);
      return null;
    }
    return item.value;
  }

  async set(key: string, value: string, ttlSeconds?: number): Promise<void> {
    const expiresAt = ttlSeconds && ttlSeconds > 0 ? Date.now() + ttlSeconds * 1000 : 0;
    this.items.set(key, { value, expiresAt });
  }

  async delete(key: string): Promise<void> {
    this.items.delete(key);
  }

  async snapshot(): Promise<string[]> {
    const now = Date.now();
    const out: string[] = [];
    for (const item of this.items.values()) {
      if (!this.isExpired(item, now)) {
        out.push(item.value);
      }
    }
    return out;
  }
}

/** Builds the appropriate store for the given env. */
export function createStore(kv: KVNamespace | undefined): Store {
  if (kv) {
    return new KVStore(kv);
  }
  return new MemoryStore();
}

File: internal/store/store.go
// Package store provides the persistence layer for fox-shield.
//
// It exposes a Store interface with an in-memory fallback implementation
// (MemoryStore) and a Redis-backed implementation (RedisStore). The same key
// schema is shared with the edge worker's KV namespace:
//
//	ratelimit:{ip}   -> sliding-window request counters
//	ban:{ip}         -> banned IPs (value = reason)
//	dark:{hash}      -> dark-listed request hashes (value = reason)
//
// If Redis is unavailable (no REDIS_URL or connection error) the shield falls
// back to MemoryStore so the origin proxy still runs.
package store

import (
	"container/list"
	"context"
	"errors"
	"strings"
	"sync"
	"time"
)

// ErrNotFound is returned when a key does not exist in the store.
var ErrNotFound = errors.New("store: key not found")

// Store is the persistence contract used by the shield middleware chain.
type Store interface {
	// Get returns the value stored under key, or ErrNotFound.
	Get(ctx context.Context, key string) (string, error)
	// Set stores value under key with an optional TTL (0 = no expiry).
	Set(ctx context.Context, key, value string, ttl time.Duration) error
	// Incr atomically increments the integer stored under key and returns the
	// new value. If the key does not exist it is created as 1.
	Incr(ctx context.Context, key string) (int64, error)
	// Expire sets a TTL on an existing key.
	Expire(ctx context.Context, key string, ttl time.Duration) error
	// Delete removes a key.
	Delete(ctx context.Context, key string) error
	// Close releases any underlying resources.
	Close() error
}

// MemoryStore is a thread-safe in-memory implementation of Store. It is used
// as a fallback when Redis is not configured or unreachable.
//
// Ban entries (keys prefixed with "ban:") are bounded by maxBans using an LRU
// eviction policy so an attacker rotating many distinct IPs cannot exhaust
// memory by creating unbounded ban records.
type MemoryStore struct {
	mu    sync.RWMutex
	items map[string]memItem
	// banLRU tracks ban keys in recency order (front = most recently used) so
	// the oldest ban can be evicted when the cap is reached.
	banLRU    *list.List
	banOrder  map[string]*list.Element
}

// maxBans caps the number of in-memory ban records to prevent memory
// exhaustion from an attacker rotating many distinct IPs.
const maxBans = 50_000

type memItem struct {
	value     string
	expiresAt time.Time
}

// NewMemoryStore returns an empty in-memory store.
func NewMemoryStore() *MemoryStore {
	return &MemoryStore{
		items:    make(map[string]memItem),
		banLRU:   list.New(),
		banOrder: make(map[string]*list.Element),
	}
}

func (m *MemoryStore) Get(_ context.Context, key string) (string, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	it, ok := m.items[key]
	if !ok {
		return "", ErrNotFound
	}
	if !it.expiresAt.IsZero() && time.Now().After(it.expiresAt) {
		delete(m.items, key)
		return "", ErrNotFound
	}
	return it.value, nil
}

func (m *MemoryStore) Set(_ context.Context, key, value string, ttl time.Duration) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	exp := time.Time{}
	if ttl > 0 {
		exp = time.Now().Add(ttl)
	}
	m.items[key] = memItem{value: value, expiresAt: exp}
	if strings.HasPrefix(key, "ban:") {
		m.touchBan(key)
	}
	return nil
}

// touchBan records a ban key as most-recently-used and evicts the oldest ban
// when the cap is exceeded. Callers must hold m.mu.
func (m *MemoryStore) touchBan(key string) {
	if el, ok := m.banOrder[key]; ok {
		m.banLRU.MoveToFront(el)
		return
	}
	el := m.banLRU.PushFront(key)
	m.banOrder[key] = el
	if m.banLRU.Len() > maxBans {
		oldest := m.banLRU.Back()
		if oldest != nil {
			oldKey := oldest.Value.(string)
			delete(m.banOrder, oldKey)
			delete(m.items, oldKey)
			m.banLRU.Remove(oldest)
		}
	}
}

func (m *MemoryStore) Incr(_ context.Context, key string) (int64, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	it, ok := m.items[key]
	if !ok || (!it.expiresAt.IsZero() && time.Now().After(it.expiresAt)) {
		m.items[key] = memItem{value: "1"}
		return 1, nil
	}
	n, err := parseCount(it.value)
	if err != nil {
		n = 0
	}
	n++
	m.items[key] = memItem{value: itoa(n), expiresAt: it.expiresAt}
	return n, nil
}

func (m *MemoryStore) Expire(_ context.Context, key string, ttl time.Duration) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	it, ok := m.items[key]
	if !ok {
		return ErrNotFound
	}
	it.expiresAt = time.Now().Add(ttl)
	m.items[key] = it
	return nil
}

func (m *MemoryStore) Delete(_ context.Context, key string) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	delete(m.items, key)
	if strings.HasPrefix(key, "ban:") {
		if el, ok := m.banOrder[key]; ok {
			m.banLRU.Remove(el)
			delete(m.banOrder, key)
		}
	}
	return nil
}

func (m *MemoryStore) Close() error { return nil }

// Snapshot returns a copy of all non-expired values currently in the store.
// It is used by the similarity detector to scan the dark list.
func (m *MemoryStore) Snapshot() []string {
	m.mu.RLock()
	defer m.mu.RUnlock()
	now := time.Now()
	out := make([]string, 0, len(m.items))
	for k, it := range m.items {
		if !it.expiresAt.IsZero() && now.After(it.expiresAt) {
			continue
		}
		_ = k
		out = append(out, it.value)
	}
	return out
}

// parseCount converts a stored string to an int64.
func parseCount(s string) (int64, error) {
	var n int64
	if len(s) == 0 {
		return 0, errors.New("store: empty count")
	}
	for i := 0; i < len(s); i++ {
		c := s[i]
		if c < '0' || c > '9' {
			return 0, errors.New("store: invalid count")
		}
		n = n*10 + int64(c-'0')
	}
	return n, nil
}

// itoa converts an int64 to a decimal string without importing strconv.
func itoa(n int64) string {
	if n == 0 {
		return "0"
	}
	neg := n < 0
	if neg {
		n = -n
	}
	var buf [20]byte
	i := len(buf)
	for n > 0 {
		i--
		buf[i] = byte('0' + n%10)
		n /= 10
	}
	if neg {
		i--
		buf[i] = '-'
	}
	return string(buf[i:])
}
